File: Core/Model/Impl/RouteEventBehaviorList.cs
namespace BinkyRailways.Core.Model.Impl
{
    /// <summary>
    /// Set of of junctions with their state.
    /// </summary>
    public sealed class RouteEventBehaviorList : EntityList<RouteEventBehavior, IRouteEventBehavior>
    {
        private readonly RouteEvent @event;

        /// <summary>
        /// Default ctor
        /// </summary>
        internal RouteEventBehaviorList(RouteEvent @event)
            : base(@event)
        {
            this.@event = @event;
        }

        /// <summary>
        /// Implementation of IJunctionWithStateSet 
        /// </summary>
        public new IRouteEventBehaviorList List
        {
            get { return (IRouteEventBehaviorList)base.List; }
        }

        /// <summary>
        /// The given item has been added to this set.
        /// </summary>
        protected override void OnAdded(RouteEventBehavior item)
        {
            item.RouteEvent = @event;
        }

        /// <summary>
        /// The given item has been removed from this set.
        /// </summary>
        protected override void OnRemoved(RouteEventBehavior item)
        {
            item.RouteEvent = null;
        }

        /// <summary>
        /// Create an implementation of IEntitySet.
        /// </summary>
        protected override ListImpl CreateListImpl()
        {
            return new ListImpl2(this);
        }

        /// <summary>
        /// Refresh the RouteEvent property of each item.
        /// </summary>
        internal void LinkItems()
        {
            foreach (var item in this)
            {
                item.RouteEvent = @event;
                item.LinkItem();
            }
        }

        private sealed class ListImpl2 : ListImpl, IRouteEventBehaviorList
        {
            private readonly RouteEventBehaviorList impl;

            /// <summary>
            /// Default ctor
            /// </summary>
            internal ListImpl2(RouteEventBehaviorList impl)
                : base(impl)
            {
                this.impl = impl;
            }

            /// <summary>
            /// Add the given item to this set
            /// </summary>
            IRouteEventBehavior IRouteEventBehaviorList.Add()
            {
                var item = new RouteEventBehavior();
                impl.Add(item);
                return item;
            }

            /// <summary>
            /// Add the given item to this set
            /// </summary>
            IRouteEventBehavior IRouteEventBehaviorList.Add(ILocPredicate appliesTo)
            {
                var item = new RouteEventBehavior { AppliesTo = (LocStandardPredicate)appliesTo };
                impl.Add(item);
                return item;
            }

            /// <summary>
            /// Remove the given item from this set.
            /// </summary>
            /// <returns>True if it was removed, false otherwise</returns>
            bool IRouteEventBehaviorList.Remove(IRouteEventBehavior item)
            {
                return impl.Remove((RouteEventBehavior) item);
            }

            /// <summary>
            /// Remove all
            /// </summary>
            void IRouteEventBehaviorList.Clear()
            {
                impl.Clear();
            }
        }
    }
}

File: pkg/core/model/impl/route_event_behavior_list.go
// Copyright 2020 Ewout Prangsma
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author Ewout Prangsma
//

package impl

import (
	"encoding/xml"

	"github.com/binkyrailways/BinkyRailways/pkg/core/model"
)

type routeEventBehaviorList struct {
	routeEventBehaviorListItems
	moduleEntityContainer
}

type routeEventBehaviorListItems struct {
	Items []*routeEventBehavior `xml:"RouteEventBehavior"`
}

var _ model.RouteEventBehaviorList = &routeEventBehaviorList{}

// UnmarshalXML unmarshals and connects the module.
func (rs *routeEventBehaviorList) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
	if err := d.DecodeElement(&rs.routeEventBehaviorListItems, &start); err != nil {
		return err
	}
	for _, x := range rs.Items {
		x.SetContainer(rs)
	}
	return nil
}

// Get number of entries
func (rs *routeEventBehaviorList) GetCount() int {
	return len(rs.Items)
}

// Invoke the callback for each item
func (rs *routeEventBehaviorList) ForEach(cb func(model.RouteEventBehavior)) {
	for _, x := range rs.Items {
		cb(x)
	}
}

// Remove the given item from this set.
// Returns true if it was removed, false otherwise
func (rs *routeEventBehaviorList) Remove(item model.RouteEventBehavior) bool {
	for i, x := range rs.Items {
		if x == item {
			rs.Items = append(rs.Items[:i], rs.Items[i+1:]...)
			return true
		}
	}
	return false
}

// Remove all items.
// Returns true if one or more items were removed, false otherwise
func (rs *routeEventBehaviorList) Clear() bool {
	if rs.GetCount() == 0 {
		return false
	}
	rs.Items = nil
	return true
}

// Add a blank route behavior to the list
func (rs *routeEventBehaviorList) AddNew() model.RouteEventBehavior {
	reb := newRouteEventBehavior()
	reb.SetContainer(rs)
	rs.Items = append(rs.Items, reb)
	rs.OnModified()
	return reb
}

/// <summary>
/// Add the given item to this set
/// </summary>
//IRouteEventBehavior Add(ILocPredicate appliesTo);
